server/devpi_server/model.py

```diff
@@ -343,35 +345,27 @@ def create_stage(self, index, type="stage", **kwargs):
             kwargs.setdefault("acl_toxresult_upload", [":ANONYMOUS:"])
             kwargs.setdefault("bases", [])
             kwargs.setdefault("mirror_whitelist", [])
-        ixconfig = get_indexconfig(
-            self.xom.config.hook,
-            type=type, **kwargs)
-        if "bases" in ixconfig:
-            ixconfig["bases"] = tuple(normalize_bases(
-                self.xom.model, ixconfig["bases"]))
-        # modify user/indexconfig
-        with self.key.update() as userconfig:
-            indexes = userconfig.setdefault("indexes", {})
-            if index in indexes:
-                raise InvalidIndex("indexname '%s' already exists" % index)
-            indexes[index] = ixconfig
-        stage = self.getstage(index)
+        stage = self._getstage(index, type, {"type": type})
+        stage._modify(**kwargs)
         threadlog.info("created index %s: %s", stage.name, stage.ixconfig)
         return stage
 
+    def _getstage(self, indexname, index_type, ixconfig):
+        if index_type == "mirror":
```

Comment:
I do not like branching on ``index_type``

Reply:
In this case it is not avoidable (yet). But the final version of this PR removed several similar cases.